File: server/internal/api/sync_test.go
package api_test

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/MohammadBnei/wird/server/internal/store"
)

// post sends a batch the way a device does: one token, one body, the answer
// read back per op.
func (h *harness) post(t *testing.T, path, token string, body any) *httptest.ResponseRecorder {
	t.Helper()
	encoded, err := json.Marshal(body)
	if err != nil {
		t.Fatalf("encode request: %v", err)
	}
	r := httptest.NewRequest(http.MethodPost, path, bytes.NewReader(encoded))
	if token != "" {
		r.Header.Set("Authorization", "Bearer "+token)
	}
	w := httptest.NewRecorder()
	h.routes.ServeHTTP(w, r)
	return w
}

// flush is one reconnect: the device hands over its queue and reads the
// verdict on each op separately.
func (h *harness) flush(t *testing.T, token string, ops ...store.Op) map[string]store.OpResult {
	t.Helper()
	w := h.post(t, "/v1/sync", token, map[string]any{"ops": ops})
	var answer struct {
		Results []store.OpResult `json:"results"`
	}
	decode(t, w, http.StatusOK, &answer)
	if len(answer.Results) != len(ops) {
		t.Fatalf("sent %d ops and got %d verdicts: a device cannot tell which of its writes landed",
			len(ops), len(answer.Results))
	}
	byID := map[string]store.OpResult{}
	for _, r := range answer.Results {
		byID[r.ClientOpID] = r
	}
	return byID
}

func (h *harness) keptList(t *testing.T, token string) []store.KeptItem {
	t.Helper()
	var answer struct {
		Items []store.KeptItem `json:"items"`
	}
	decode(t, h.get(t, "/v1/kept", token), http.StatusOK, &answer)
	return answer.Items
}

func (h *harness) pull(t *testing.T, token, cursor string) store.Changes {
	t.Helper()
	var page store.Changes
	decode(t, h.get(t, "/v1/changes?since="+cursor, token), http.StatusOK, &page)
	return page
}

func op(id, kind string, body any) store.Op {
	encoded, err := json.Marshal(body)
	if err != nil {
		panic(err)
	}
	return store.Op{ClientOpID: id, Kind: kind, Body: encoded}
}

// opID makes a uuid out of a number so a test can read which op is which.
func opID(n int) string {
	return fmt.Sprintf("00000000-0000-4000-8000-%012d", n)
}

var noon = time.Date(2026, 3, 1, 12, 0, 0, 0, time.UTC)

func aSet(id string, ordinal int) store.Op {
	return op(id, "set_recorded", map[string]any{
		"id": id, "ordinal": ordinal, "start_ayah_id": 96001, "end_ayah_id": 96005,
		"reading_order": "nuzul", "created_at": noon,
	})
}

// The failure: a flush that timed out after the server had already applied it
// is sent again on the next reconnect, and the reader's prayer count goes up
// twice for one prayer. That number is the one the app exists to show, and it
// is never corrected.
func TestAReplayedFlushCountsOnePrayerNotTwo(t *testing.T) {
	h := newHarness(t)
	token := h.tokenFor(t, "sub-replay")

	setID := "11111111-1111-4111-8111-111111111111"
	prayer := op(opID(2), "set_prayed", map[string]any{
		"id": "22222222-2222-4222-8222-222222222222", "set_id": setID,
		"prayer_name": "fajr", "prayed_at": noon,
	})
	understood := op(opID(3), "ayah_understood", map[string]any{
		"ayah_ids": []int{96001, 96002, 96003}, "understood_at": noon,
	})
	batch := []store.Op{aSet(setID, 1), prayer, understood}

	first := h.flush(t, token, batch...)
	for _, o := range batch {
		if first[o.ClientOpID].Status != store.OpApplied {
			t.Fatalf("first flush of %s: %+v", o.Kind, first[o.ClientOpID])
		}
	}

	// The same queue, twice more: the phone reconnected, lost the answer, and
	// tried again.
	for attempt := 2; attempt <= 3; attempt++ {
		again := h.flush(t, token, batch...)
		for _, o := range batch {
			if again[o.ClientOpID].Status != store.OpDuplicate {
				t.Fatalf("reconnect %d applied %s a second time: %+v", attempt, o.Kind, again[o.ClientOpID])
			}
		}
	}

	var p store.Progress
	decode(t, h.get(t, "/v1/progress", token), http.StatusOK, &p)
	if p.Prayers != 1 {
		t.Fatalf("one prayer prayed, %d counted", p.Prayers)
	}
	if p.Sets != 1 {
		t.Fatalf("one set read, %d counted", p.Sets)
	}
	if p.Understood != 3 {
		t.Fatalf("three ayas understood, %d counted", p.Understood)
	}
}

// The failure: a flush is cut off halfway — the socket dies, the server
// answers 500 after ten of twenty ops. The device resends the whole queue.
// The ten that landed must not land again, and the ten that did not must land.
func TestAFlushCutOffHalfwayFinishesOnTheRetryWithoutDoublingWhatLanded(t *testing.T) {
	h := newHarness(t)
	token := h.tokenFor(t, "sub-halfway")

	setID := "33333333-3333-4333-8333-333333333333"
	h.flush(t, token, aSet(setID, 1))

	prayers := make([]store.Op, 20)
	for i := range prayers {
		prayers[i] = op(opID(100+i), "set_prayed", map[string]any{
			"id":          fmt.Sprintf("44444444-4444-4444-8444-%012d", i),
			"set_id":      setID,
			"prayer_name": "isha",
			"prayed_at":   noon.Add(time.Duration(i) * time.Minute),
		})
	}

	// The half the server got to before the connection died.
	h.flush(t, token, prayers[:10]...)
	// The device kept all twenty, because it never read an answer.
	retry := h.flush(t, token, prayers...)

	for i, o := range prayers {
		want := store.OpApplied
		if i < 10 {
			want = store.OpDuplicate
		}
		if retry[o.ClientOpID].Status != want {
			t.Fatalf("prayer %d came back %+v, wanted %s", i, retry[o.ClientOpID], want)
		}
	}

	var p store.Progress
	decode(t, h.get(t, "/v1/progress", token), http.StatusOK, &p)
	if p.Prayers != 20 {
		t.Fatalf("twenty prayers prayed, %d counted", p.Prayers)
	}
}

// The failure the plan names: one op the server will never accept sits at the
// head of the queue and holds the other nineteen behind it on every reconnect,
// so the reader silently loses everything they did.
func TestAPoisonOpDoesNotBlockTheNineteenWritesBehindIt(t *testing.T) {
	h := newHarness(t)
	token := h.tokenFor(t, "sub-poison")

	// A prayer against a set this account does not have — the plan's own
	// example of a poison op.
	poison := op(opID(1), "set_prayed", map[string]any{
		"id":          "55555555-5555-4555-8555-555555555555",
		"set_id":      "99999999-9999-4999-8999-999999999999",
		"prayer_name": "maghrib",
		"prayed_at":   noon,
	})

	ops := []store.Op{poison}
	for i := range 19 {
		ops = append(ops, op(opID(200+i), "ayah_understood", map[string]any{
			"ayah_ids": []int{96001 + i}, "understood_at": noon,
		}))
	}

	results := h.flush(t, token, ops...)
	if got := results[poison.ClientOpID]; got.Status != store.OpRefused {
		t.Fatalf("the poison op came back %+v; a device cannot tell it to stop resending", got)
	}
	if results[poison.ClientOpID].Reason == "" {
		t.Fatal("the refusal carries no reason, so settings has nothing to show the reader")
	}
	for _, o := range ops[1:] {
		if results[o.ClientOpID].Status != store.OpApplied {
			t.Fatalf("a good write behind the poison op came back %+v", results[o.ClientOpID])
		}
	}

	var p store.Progress
	decode(t, h.get(t, "/v1/progress", token), http.StatusOK, &p)
	if p.Understood != 19 {
		t.Fatalf("nineteen writes were queued behind one bad one and %d landed", p.Understood)
	}

	// And it stays refused however often the device tries, which is what makes
	// the client's five attempts terminate instead of looping forever.
	for attempt := 2; attempt <= 5; attempt++ {
		if got := h.flush(t, token, poison)[poison.ClientOpID]; got.Status != store.OpRefused {
			t.Fatalf("attempt %d of the poison op came back %+v", attempt, got)
		}
	}
}

// The failure: the reader deletes a note on the phone and the tablet hands it
// straight back on the next pull, forever. A delete has to travel as a row,
// not as an absence.
func TestAKeptItemDeletedOnThePhoneStaysDeletedOnTheTablet(t *testing.T) {
	h := newHarness(t)
	token := h.tokenFor(t, "sub-two-devices")
	const note = "66666666-6666-4666-8666-666666666666"

	// The phone keeps a note and flushes.
	h.flush(t, token, op(opID(1), "kept_upsert", map[string]any{
		"id": note, "kind": "note", "ayah_id": 96001, "root_letters": nil,
		"body": "the first word revealed", "tags": []string{"revisit"},
		"created_at": noon, "updated_at": noon,
	}))

	// The tablet pulls and holds it.
	tablet := h.pull(t, token, "")
	if !holds(tablet, note) {
		t.Fatalf("the tablet never received the note: %+v", tablet.Changes)
	}

	// The phone deletes it and flushes.
	h.flush(t, token, op(opID(2), "kept_delete", map[string]any{
		"id": note, "deleted_at": noon.Add(time.Hour),
	}))

	// The tablet pulls again from where it left off.
	after := h.pull(t, token, tablet.Cursor)
	tombstone, found := change(after, note)
	if !found {
		t.Fatal("the delete never reached the tablet, so the note is still on its list")
	}
	var row struct {
		DeletedAt *time.Time `json:"deleted_at"`
	}
	if err := json.Unmarshal(tombstone.Row, &row); err != nil {
		t.Fatalf("tombstone row: %v", err)
	}
	if row.DeletedAt == nil {
		t.Fatalf("the row came back alive: %s", tombstone.Row)
	}

	// And the server's own read model agrees, so a third device that syncs
	// from scratch does not resurrect it either.
	kept := h.keptList(t, token)
	if len(kept) != 0 {
		t.Fatalf("a deleted note is still on the kept list: %+v", kept)
	}
}

// The failure: the tablet asks for what changed and is handed the reader's
// whole history again, every time, because the cursor it was given does not
// move.
func TestASecondPullDoesNotHandBackWhatTheDeviceAlreadyHas(t *testing.T) {
	h := newHarness(t)
	token := h.tokenFor(t, "sub-cursor")

	h.flush(t, token, op(opID(1), "ayah_understood", map[string]any{
		"ayah_ids": []int{96001, 96002}, "understood_at": noon,
	}))

	first := h.pull(t, token, "")
	if len(first.Changes) != 2 {
		t.Fatalf("first pull carried %d rows, expected the two ayas", len(first.Changes))
	}
	if first.Cursor == "" {
		t.Fatal("no cursor came back, so the next pull must start from the beginning")
	}

	if again := h.pull(t, token, first.Cursor); len(again.Changes) != 0 {
		t.Fatalf("the same rows came back a second time: %+v", again.Changes)
	}

	// One new write, and only that one arrives.
	h.flush(t, token, op(opID(2), "ayah_understood", map[string]any{
		"ayah_ids": []int{96003}, "understood_at": noon.Add(time.Hour),
	}))
	third := h.pull(t, token, first.Cursor)
	if len(third.Changes) != 1 {
		t.Fatalf("the incremental pull carried %d rows, expected one: %+v", len(third.Changes), third.Changes)
	}
}

// The failure: one reader's device sends an op naming another reader's kept
// item and edits or deletes it. Ids are minted on devices, so a guessed id
// must not be authority over somebody else's note.
func TestOneReadersOpCannotTouchAnotherReadersNote(t *testing.T) {
	h := newHarness(t)
	owner := h.tokenFor(t, "sub-owner")
	stranger := h.tokenFor(t, "sub-stranger")
	const note = "77777777-7777-4777-8777-777777777777"

	h.flush(t, owner, op(opID(1), "kept_upsert", map[string]any{
		"id": note, "kind": "note", "ayah_id": nil, "root_letters": nil,
		"body": "mine", "tags": []string{}, "created_at": noon, "updated_at": noon,
	}))

	// The stranger tries to overwrite it, then to delete it.
	h.flush(t, stranger, op(opID(2), "kept_upsert", map[string]any{
		"id": note, "kind": "note", "ayah_id": nil, "root_letters": nil,
		"body": "theirs", "tags": []string{}, "created_at": noon,
		"updated_at": noon.Add(time.Hour),
	}))
	h.flush(t, stranger, op(opID(3), "kept_delete", map[string]any{
		"id": note, "deleted_at": noon.Add(2 * time.Hour),
	}))

	kept := h.keptList(t, owner)
	if len(kept) != 1 || kept[0].Body != "mine" {
		t.Fatalf("a stranger reached the owner's note: %+v", kept)
	}

	theirs := h.keptList(t, stranger)
	if len(theirs) != 0 {
		t.Fatalf("the stranger's write landed on their own account instead of being refused: %+v", theirs)
	}
}

// The failure: an older phone sends a body the server cannot read and the
// server takes the whole batch down with a 500, so the device retries the same
// unreadable queue on every reconnect.
func TestAnOpTheServerCannotReadIsRefusedAloneAndNotAsA500(t *testing.T) {
	h := newHarness(t)
	token := h.tokenFor(t, "sub-garbage")

	results := h.flush(t, token,
		store.Op{ClientOpID: opID(1), Kind: "ayah_understood", Body: json.RawMessage(`{"ayah_ids":"all of them"}`)},
		store.Op{ClientOpID: opID(2), Kind: "a kind this server has never heard of", Body: json.RawMessage(`{}`)},
		store.Op{ClientOpID: opID(3), Kind: "ayah_understood", Body: json.RawMessage(`{"ayah_ids":[115001],"understood_at":"2026-03-01T12:00:00Z"}`)},
		op(opID(4), "ayah_understood", map[string]any{"ayah_ids": []int{96001}, "understood_at": noon}),
	)

	for _, bad := range []int{1, 2, 3} {
		if got := results[opID(bad)]; got.Status != store.OpRefused {
			t.Fatalf("op %d came back %+v, so the device will resend it forever", bad, got)
		}
	}
	if results[opID(4)].Status != store.OpApplied {
		t.Fatalf("the good op behind three bad ones came back %+v", results[opID(4)])
	}

	// A body that is not a batch at all is the one thing refused whole — the
	// device has nothing per-op to be told about.
	if code := h.post(t, "/v1/sync", token, "not a batch").Code; code != http.StatusBadRequest {
		t.Fatalf("an unreadable batch answered %d", code)
	}
}

// The failure: a cursor the device made up, or one left over from a server
// that issued a different shape, silently reads as "the beginning" and the
// reader's whole history arrives as new.
func TestACursorThisServerNeverIssuedIsRefusedRatherThanReadAsTheBeginning(t *testing.T) {
	h := newHarness(t)
	token := h.tokenFor(t, "sub-bad-cursor")

	for _, cursor := range []string{"lastweek", "42", "2026-03-01T12:00:00Z"} {
		if code := h.get(t, "/v1/changes?since="+cursor, token).Code; code != http.StatusBadRequest {
			t.Errorf("cursor %q answered %d, expected 400", cursor, code)
		}
	}
}

func change(page store.Changes, id string) (store.Change, bool) {
	for _, c := range page.Changes {
		if c.ID == id {
			return c, true
		}
	}
	return store.Change{}, false
}

func holds(page store.Changes, id string) bool {
	_, found := change(page, id)
	return found
}

File: app/test/data/sync_test.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:sqflite_common_ffi/sqflite_ffi.dart';
import 'package:wird/data/db.dart';
import 'package:wird/data/kept_repo.dart';
import 'package:wird/data/outbox.dart';
import 'package:wird/data/sets.dart';
import 'package:wird/data/sync.dart';

import '../corpus.dart';
import 'fake_wird.dart';

Map<String, dynamic> keptRow(
  String id, {
  required String updatedAt,
  String body = '',
  String? deletedAt,
}) => {
  'kind': 'kept_items',
  'id': id,
  'at': updatedAt,
  'row': {
    'id': id,
    'kind': 'note',
    'ayah_id': null,
    'root_letters': null,
    'body': body,
    'tags': <String>[],
    'created_at': updatedAt,
    'updated_at': updatedAt,
    'deleted_at': deletedAt,
  },
};

Future<int> queued(Database db) async =>
    (await db.rawQuery('SELECT COUNT(*) AS n FROM outbox')).single['n']! as int;

/// The wait a transient failure earned, served. Real time cannot be waited out
/// in a test, and a fake clock seam would exist only for the test to hold.
Future<void> theClockMovesOn(Database db) =>
    db.rawUpdate('UPDATE outbox SET retry_after = NULL');

void main() {
  late Database db;
  late FakeWird server;

  setUp(() async {
    db = await testCorpus();
    await ensureKeptTable(db);
    await db.delete('kept_items');
    await db.execute('DROP TABLE IF EXISTS sync_state');
    server = await FakeWird.start();
  });

  tearDown(() async => server.stop());

  // The offline promise, two sets deep. Everything the reader did at 30,000
  // feet reaches the server when the plane lands — once each, not twice, and
  // not never.
  test('two sets read in airplane mode land exactly once on reconnect', () async {
    await server.stop(); // the plane

    await markSetUnderstood(db, newOpId(), [96001, 96002, 96003]);
    await keep(db, kind: KeptKind.note, body: 'the pen and what it writes');
    await markSetUnderstood(db, newOpId(), [68001, 68002]);
    await setReadingOrder(db, ReadingOrder.mushaf);

    final offline = await syncNow(db, SyncApi(server.dio));
    expect(offline.reachedServer, isFalse);
    expect(
      await queued(db),
      4,
      reason: 'a flush with no server to reach must not drop the queue',
    );

    // The plane lands.
    server = await FakeWird.start();
    final landed = await syncNow(db, SyncApi(server.dio));

    expect(landed.reachedServer, isTrue);
    expect(landed.landed, 4);
    expect(await queued(db), 0, reason: 'the queue did not drain');

    final kinds = server.opsReceived.map((op) => op['kind']).toList();
    expect(kinds, [
      'ayah_understood',
      'kept_upsert',
      'ayah_understood',
      'prefs_set',
    ]);
    final ids = server.opsReceived.map((op) => op['client_op_id']).toSet();
    expect(ids, hasLength(4), reason: 'an op was sent twice in one flush');
    // The reader's second set is still understood locally, whatever the
    // network did.
    expect(
      (await db.rawQuery(
        'SELECT COUNT(*) AS n FROM ayah_understood',
      )).single['n'],
      5,
    );
  });

  // The failure: a flush lands on the server and the answer is lost on the way
  // back. If the device dropped the op on send, the write is gone; if it
  // dropped it without an answer, a prayer is counted twice. It must send the
  // same op id again and wait to be told.
  test('a flush whose answer never arrived is sent again under the same op id, '
      'so the prayer is counted once', () async {
    final opId = newOpId();
    await markSetUnderstood(db, opId, [96001, 96002, 96003]);

    // The answer never comes back.
    await server.stop();
    await syncNow(db, SyncApi(server.dio));
    expect(await queued(db), 1);

    // The reconnect. The server has it already and says so.
    server = await FakeWird.start();
    server.verdict = (_) => 'duplicate';
    final report = await syncNow(db, SyncApi(server.dio));

    expect(server.opsReceived.single['client_op_id'], opId);
    expect(report.landed, 1);
    expect(
      await queued(db),
      0,
      reason: 'an op the server already has must leave the queue',
    );
  });

  // The plan's own defect: one op the server will never accept sits at the
  // head of the queue and holds everything behind it, on every reconnect,
  // until the reader loses it all.
  test('a poison op does not block the nineteen writes queued behind it', () async {
    final poison = newOpId();
    await markSetUnderstood(db, poison, [96001]);
    for (var i = 0; i < 19; i++) {
      await markSetUnderstood(db, newOpId(), [68001 + i]);
    }
    server.verdict = (id) => id == poison ? 'refused' : 'applied';

    final report = await syncNow(db, SyncApi(server.dio));

    expect(report.landed, 19);
    expect(report.refused, 1);
    expect(await pending(db), isEmpty);
    final parked = await deadLettered(db);
    expect(
      parked.map((op) => op.id),
      [poison],
      reason: 'the refused op must leave the queue, not sit at its head',
    );
  });

  // The server's own word for a refusal is that it "will never succeed however
  // often it is sent". Asking it four more times delays the moment the reader
  // is told and changes nothing else.
  test('a refused write is asked about four more times before the reader hears '
      'about it', () async {
    final poison = newOpId();
    await markSetUnderstood(db, poison, [96001]);
    server.verdict = (_) => 'refused';

    await syncNow(db, SyncApi(server.dio));

    final dead = await deadLettered(db);
    expect(dead.map((op) => op.id), [poison]);
    expect(dead.single.body['ayah_ids'], [96001]);

    // A good write made afterwards still flushes; the parked one does not ride
    // with it.
    server.verdict = (_) => 'applied';
    await markSetUnderstood(db, newOpId(), [68001]);
    await syncNow(db, SyncApi(server.dio));

    expect(
      server.batches.last.map((op) => op['client_op_id']),
      isNot(contains(poison)),
    );
  });

  // The defect: the server says `failed` and means "not applied, try again",
  // and the device spent one of five lives on it anyway. Five bad minutes
  // across five reconnects and a write nothing was wrong with is parked for
  // good.
  test('five transient server faults park a good write that should still be '
      'trying', () async {
    final good = newOpId();
    await markSetUnderstood(db, good, [96001]);
    server.verdict = (_) => 'failed';

    for (var reconnect = 0; reconnect < 5; reconnect++) {
      final report = await syncNow(db, SyncApi(server.dio));
      expect(report.reachedServer, isTrue);
      await theClockMovesOn(db);
    }

    expect(
      await deadLettered(db),
      isEmpty,
      reason: 'the server asked to be tried again, and was answering each time',
    );

    // The server has its good minute back.
    server.verdict = (_) => 'applied';
    final landed = await syncNow(db, SyncApi(server.dio));
    expect(landed.landed, 1);
    expect(await queued(db), 0);
  });

  // Without the delay the count is not a budget: a phone that reconnects every
  // few seconds — a train, a lift, a flaky router — spends all ten answers
  // before the server has finished restarting.
  test('a phone reconnecting every few seconds spends the whole retry budget '
      'during one restart', () async {
    await markSetUnderstood(db, newOpId(), [96001]);
    server.verdict = (_) => 'failed';

    await syncNow(db, SyncApi(server.dio));
    await syncNow(db, SyncApi(server.dio));
    await syncNow(db, SyncApi(server.dio));

    expect(
      server.batches,
      hasLength(1),
      reason: 'the second and third reconnects arrived inside the backoff',
    );
    expect(await deadLettered(db), isEmpty);
  });

  // And the budget is a budget: a server that keeps failing for good does not
  // leave the write pending forever with nobody told.
  test('a server that never recovers leaves the write pending forever and '
      'nobody is told', () async {
    await markSetUnderstood(db, newOpId(), [96001]);
    server.verdict = (_) => 'failed';

    for (var reconnect = 0; reconnect < maxAttempts; reconnect++) {
      await syncNow(db, SyncApi(server.dio));
      await theClockMovesOn(db);
    }

    expect(await pending(db), isEmpty);
    expect(await deadLettered(db), hasLength(1));
  });

  // The failure: a reader spends a week somewhere with no signal, the flush
  // fails five times against nothing at all, and the queue is thrown away
  // before it ever reaches a server.
  test('a week with no signal does not dead-letter a single write', () async {
    await markSetUnderstood(db, newOpId(), [96001]);
    await server.stop();

    for (var day = 0; day < 7; day++) {
      final report = await syncNow(db, SyncApi(server.dio));
      expect(report.reachedServer, isFalse);
    }

    expect(await deadLettered(db), isEmpty);
    expect((await pending(db)).single.attempts, 0);
  });

  // The failure the plan names: without the tombstone as a row, a note deleted
  // on the phone is handed straight back by the tablet on every pull.
  test('a kept note deleted on the phone is not resurrected by the tablet', () async {
    final id = await keep(db, kind: KeptKind.note, body: 'a thought');
    await forget(db, id);
    await syncNow(db, SyncApi(server.dio));

    // The tablet's copy of the note comes back on the next pull, as it will:
    // it is the same row, carrying the delete.
    final deletedAt = DateTime.now().toUtc().toIso8601String();
    server.pages = [
      {
        'changes': [
          keptRow(id, updatedAt: deletedAt, body: 'a thought', deletedAt: deletedAt),
        ],
        'cursor': '$deletedAt|$id',
        'more': false,
      },
    ];
    await syncNow(db, SyncApi(server.dio));

    expect(await keptItems(db), isEmpty);
    final row = (await db.query(
      'kept_items',
      where: 'id = ?',
      whereArgs: [id],
    )).single;
    expect(row['deleted_at'], isNotNull);
  });

  // And the same row arriving alive, from a device that had not seen the
  // delete yet, must not undo it either.
  test('an older copy of a note from a stale device does not undo the delete',
      () async {
    final id = await keep(db, kind: KeptKind.note, body: 'a thought');
    await forget(db, id);

    final stale = DateTime.now()
        .toUtc()
        .subtract(const Duration(days: 1))
        .toIso8601String();
    server.pages = [
      {
        'changes': [keptRow(id, updatedAt: stale, body: 'a thought')],
        'cursor': '$stale|$id',
        'more': false,
      },
    ];
    await syncNow(db, SyncApi(server.dio));

    expect(await keptItems(db), isEmpty);
  });

  // The failure: the cursor never moves, so every reconnect re-applies the
  // reader's whole history — years of it, over a phone connection.
  test('the second pull asks from the cursor and does not re-apply the whole '
      'history', () async {
    final at = DateTime.now().toUtc().toIso8601String();
    const note = '9b7c1d2e-0000-4000-8000-000000000001';
    server.pages = [
      {
        'changes': [keptRow(note, updatedAt: at, body: 'from the tablet')],
        'cursor': 'cursor-1',
        'more': false,
      },
    ];

    final first = await syncNow(db, SyncApi(server.dio));
    expect(first.applied, 1);
    expect((await keptItems(db)).single.body, 'from the tablet');

    final second = await syncNow(db, SyncApi(server.dio));
    expect(second.applied, 0);
    expect(server.cursorsAsked, ['', 'cursor-1']);
  });

  // A page that says there is more must be followed, or the reader restoring
  // onto a new phone gets the first five hundred rows and no more.
  test('a paged pull keeps asking until the server says it is finished', () async {
    final at = DateTime.now().toUtc().toIso8601String();
    server.pages = [
      {
        'changes': [keptRow('9b7c1d2e-0000-4000-8000-000000000002', updatedAt: at)],
        'cursor': 'cursor-1',
        'more': true,
      },
      {
        'changes': [keptRow('9b7c1d2e-0000-4000-8000-000000000003', updatedAt: at)],
        'cursor': 'cursor-2',
        'more': false,
      },
    ];

    final report = await syncNow(db, SyncApi(server.dio));

    expect(report.applied, 2);
    expect(server.cursorsAsked, ['', 'cursor-1']);
    expect(await keptItems(db), hasLength(2));
  });

  // The failure: hotel wifi answers every request with its own sign-in page
  // and a 200 on it. The device read that as the contract's JSON, and the
  // TypeError went straight out of the flush — past the one catch that knows
  // the queue is intact, and into whatever asked for the flush.
  test('a captive portal answering 200 throws out of the flush instead of '
      'being read as a server that was never reached', () async {
    await markSetUnderstood(db, newOpId(), [96001, 96002, 96003]);
    server.insteadAPortal = '<html><body>Sign in to Hotel Wifi</body></html>';

    final report = await syncNow(db, SyncApi(server.dio));

    expect(report.reachedServer, isFalse);
    expect(report.deadLettered, 0);
    expect(
      await queued(db),
      1,
      reason: 'the portal is not the server, so the write is still waiting',
    );

    // And the flush after it, on a network with a real server behind it,
    // carries the same write.
    server.insteadAPortal = null;
    expect((await syncNow(db, SyncApi(server.dio))).landed, 1);
  });

  // The in-prayer rule, as code rather than as a promise: the write path is
  // local. A server that accepts the connection and then says nothing at all
  // is the worst case for a screen that must not block, and the reader's mark
  // does not touch it.
  test('marking a set understood never waits on the network', () async {
    final silent = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    // Accept and answer nothing, forever.
    unawaited(silent.forEach((_) {}));
    addTearDown(() => silent.close(force: true));

    await markSetUnderstood(
      db,
      newOpId(),
      [96001, 96002, 96003],
    ).timeout(const Duration(seconds: 5));

    expect(await queued(db), 1);
  });
}
